File: src/main/scala/com/vc/hard/L37.scala
package com.vc.hard

object L37 {
  def solveSudoku(board: Array[Array[Char]]): Unit = {
    def isValid(): Boolean = {
      import scala.collection.mutable
      val set = new mutable.HashSet[String]()
      (0 until 9).foreach(i => {
        (0 until 9).foreach(j => {
          val ch = board(i)(j)
          if(ch != '.') {
            if(
              !set.add(s"$ch in ROW $i") ||
                !set.add(s"$ch in COL $j") ||
                !set.add(s"$ch in BLOCK $i/3 & $j/3")
            )
              return false
          }
        })
      })
      return true
    }
    def solve(): Boolean = {
      (0 until 9).foreach(i => {
        (0 until 9).foreach(j => {
          if(board(i)(j) == '.') {
            ('1' to '9').foreach(num => {
              board(i)(j) = num
              if(isValid()) {
                if(solve()) return true
                else return false
              }
              else board(i)(j) = '.'
            })
            return false
          }
        })
      })
      true
    }
    solve()
  }

  def main(args: Array[String]): Unit = {
    solveSudoku(
      Array(
        Array('5','3','.','.','7','.','.','.','.'),
        Array('6','.','.','1','9','5','.','.','.'),
        Array('.','9','8','.','.','.','.','6','.'),
        Array('8','.','.','.','6','.','.','.','3'),
        Array('4','.','.','8','.','3','.','.','1'),
        Array('7','.','.','.','2','.','.','.','6'),
        Array('.','6','.','.','.','.','2','8','.'),
        Array('.','.','.','4','1','9','.','.','5'),
        Array('.','.','.','.','8','.','.','7','9')
      )
    )
  }
}

File: src/main/java/com/vc/hard/L37.java
package com.vc.hard;

class L37 {
    public void solveSudoku(char[][] board) {
        solve(board);
    }

    private boolean solve(char[][] board) {
        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                if(board[i][j] == '.') {
                    for(char k = '1'; k <= '9'; k++) {
                        if(isValid(board, i, j, k)) {
                            board[i][j] = k;
                            if(solve(board)) return true;
                            else board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    private boolean isValid(char[][] board, int row, int col, char num) {
        int blkrow = (row / 3) * 3, blkcol = (col / 3) * 3;
        for(int i = 0; i < 9; i++) {
            if(board[i][col] == num || board[row][i] == num || board[blkrow + i / 3][blkcol + i % 3] == num)
                return false;
        }
        return true;
    }
}